### PR TITLE
Support nested updates of objects through `realm.create` via `UpdateMode.All`

### LINKED ISSUE #5933: UpdateMode doesn't get passed recursively when creating child objects in realm v12
### How frequently does the bug occur?

Always

### Description

Realm v12.0.0-alpha.2 always use `UpdateMode.Never` for children objects ignoring mode arg in create function:

https://github.com/realm/realm-js/blob/d357fafbd53f8a57b3a00a34b2db7d98b38b09a4/packages/realm/src/TypeHelpers.ts#L231-L238

### Stacktrace & log output

_No response_

### Can you reproduce the bug?

Always

### Reproduction Steps

Code:
```js
import Realm, { UpdateMode } from "realm";

const ASchema = {
  name: "A",
  primaryKey: "id",
  properties: {
    id: "int",
  },
};

const BSchema = {
  name: "B",
  properties: {
    a: "A[]",
  },
};

const realm = new Realm({
  schema: [ASchema, BSchema],
  schemaVersion: 0,
  inMemory: true,
  path: "temp_db",
});

realm.write(() => {
  realm.create("A", {
    id: 1,
  });
  realm.create("B", {
    a: [{ id: 1 }],
  }, UpdateMode.All);
});
```

### Expected
Successful creation

### Actual
`Error: Attempting to create an object of type 'A' with an existing primary key value '1'., js engine: hermes`

### Version

12.0.0-alpha.2

### What services are you using?

Local Database only

### Are you using encryption?

No

### Platform OS and version(s)

React-native; Android, IOS

### Build environment

Which debugger for React Native: ..

### Cocoapods version

_No response_

### PR DESCRIPTION
**Pull request by kraenhansen**
## What, How & Why?

This PR implements an initial solution for applying the `UpdateMode` to all objects when passed to `Realm.create()`.

This closes #5933.

## ☑️ ToDos
* [x] 📝 Changelog entry
* [x] 🚦 Tests

### CODE REVIEW

File: packages/realm/src/Realm.ts [hunk 1]
```diff
@@ -857,11 +859,15 @@ export class Realm {
...
-      throw new Error("Unsupported 'updateMode'. Only 'never', 'modified' or 'all' is supported.");
+      throw new Error(
+        `Unsupported 'updateMode'. Only '${UpdateMode.Never}', '${UpdateMode.Modified}' or '${UpdateMode.All}' is supported.`,
+      );
     }
     this.internal.verifyOpen();
     const helpers = this.classes.getHelpers(type);
+    this.currentUpdateMode = mode;
     const realmObject = RealmObject.create(this, values, mode, { helpers });
+    this.currentUpdateMode = undefined;
```
Comment (kraenhansen) [inline packages/realm/src/Realm.ts:870]: This will never happen if `RealmObject.create` throws.

**Comment by kraenhansen** [comment]
While having a state (`currentUpdateMode` on `Realm`) might work for now, I'd suggest avoiding storing and maintaining state for operations as it has a couple of drawbacks:
- It can cause weird bugs if not properly set between calls of the operation.
  - If the reset happen at the end, it might not happen if an exception is throw.
  - If it happens in the beginning (as with your suggestion), the deep dependent code has implicit dependencies on the state being sat before it's called, which isn't easily expressed in code and you'd need to rely on documentation.
- It will make it more difficult to transition the implementation into asynchronous code (i.e. make it harder to change if we wanted to), since that might introduce concurrent calls into the deep code and it can't reasonably expect the state to originate from its particular invocation.

All in all, I'd suggest refactoring to passing the update mode through argument, possibly by adding a property to the `CreationContext`.

**Comment by elle-j** [comment]
@kraenhansen, very valuable input! We've discussed various issues and options on Slack (see our channel) with your input in mind. For now we've opted to move forward with the initial solution (but using `try-finally` for setting/unsetting `currentUpdateMode`), and creating a ticket for finding a more ideal solution for a later v12 release.

File: packages/realm/src/TypeHelpers.ts [hunk 2]
```diff
@@ -73,7 +72,7 @@ export function toArrayBuffer(value: unknown, stringToBase64 = true) {
 
 /** @internal */
 export type TypeHelpers<T = unknown> = {
-  toBinding(value: T, createObj?: ObjCreator): binding.MixedArg;
+  toBinding(value: T, options?: { createObj?: ObjCreator; updateMode?: UpdateMode }): binding.MixedArg;
```
Comment (takameyer) [inline packages/realm/src/TypeHelpers.ts:75]: Is the `updateMode` being set somewhere?  I haven't yet seen this in any of the `toBinding` calls.

File: integration-tests/tests/src/tests/objects.ts [hunk 3]
```diff
@@ -586,6 +606,113 @@ describe("Realm.Object", () => {
...
+                    },
+                  },
+                },
+              },
+              Realm.UpdateMode.All,
+            );
+          });
+        });
+
+        it("can be updated recursively in sets", function (this: Mocha.Context & RealmContext) {
```
Comment (takameyer) [inline integration-tests/tests/src/tests/objects.ts:691]: That's amazing that this works for Sets and Dictionaries now.  
Should we also add assertions for the updated values?  We are just testing that this doesn't throw, should we also test that it does throw?

**Review by takameyer — APPROVED** [review]
Great!  Just a few clarifying questions/suggestions

File: packages/realm/src/TypeHelpers.ts (hunk 2 above)
Comment (elle-j) [reply packages/realm/src/TypeHelpers.ts:75]: Not yet, but that's the intention for the more ideal solution down the line. So I kept that refactor. (I'm mentioning it in [this TODO comment](https://github.com/realm/realm-js/blob/5e2caccff9c805205939bb4932e77d44eb6748f8/packages/realm/src/TypeHelpers.ts#L264).) Will also create a ticket and link to the ticket in that comment.

File: integration-tests/tests/src/tests/objects.ts (hunk 3 above)
Comment (takameyer) [reply integration-tests/tests/src/tests/objects.ts:713]: Great!  Thanks 🎉

File: packages/realm/src/TypeHelpers.ts (hunk 2 above)
Comment (takameyer) [reply packages/realm/src/TypeHelpers.ts:75]: Perfect 🙏🏼

File: packages/realm/src/Realm.ts (hunk 1 above)
Comment (elle-j) [reply packages/realm/src/Realm.ts]: This has been updated to use `finally`. Also, see the PR comment further below addressing your longer comment 👍

**Review by kneth — APPROVED** [review]

**Pull request merged**